### PR TITLE
Introduce parser memoization to avoid exponentional behavior

### PR DESCRIPTION
**Pull request by Kha**
Cache invocations of category, `leading_parser` and `syntax :=` parsers using a mapping from parser name and current context & state to parse result. Abstract parser context and syntax stack to enforce correct use.

### CODE REVIEW

**Comment by Kha** [comment]
!bench

**Comment by Kha** [comment]
Instruction counts suggest these are random fluctuations. Which seem to have become more frequent recently.

**Comment by Kha** [comment]
!bench

**Comment by Kha** [comment]
Comparison with previous benchmark: http://speedcenter.informatik.kit.edu/velcom/compare/6b556c5e-941a-48e3-8f57-5457ee362ab8/to/30d3b0d6-b0c2-411e-af98-c57c0f36db58 +14% parsing, though variance is high

**Comment by Kha** [comment]
!bench

**Comment by Kha** [comment]
+71% parsing compared to previous run http://speedcenter.informatik.kit.edu/velcom/compare/30d3b0d6-b0c2-411e-af98-c57c0f36db58/to/33767806-2bc2-4029-ad5f-f09a55692675

**Comment by Kha** [comment]
!bench

File: src/Lean/Parser/Types.lean [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+
+def adaptCacheableContext (f : CacheableParserContext → CacheableParserContext) : Parser → Parser :=
+  withFn (adaptCacheableContextFn f)
+
+/-- Run `p` under the given context transformation with a fresh cache, restore outer cache afterwards. -/
+def adaptUncacheableContextFn (f : ParserContextCore → ParserContextCore) (p : ParserFn) : ParserFn := fun c s =>
+  -- extract and temporarily reset parser cache
+  let ⟨stack, lhsPrec, pos, ⟨tkCache, pCache⟩, errorMsg⟩ := s
+  let ⟨stack, lhsPrec, pos, ⟨tkCache, _⟩, errorMsg⟩ := p ⟨f c.toParserContextCore⟩ ⟨stack, lhsPrec, pos, ⟨tkCache, {}⟩, errorMsg⟩
+  ⟨stack, lhsPrec, pos, ⟨tkCache, pCache⟩, errorMsg⟩
```
Comment (Kha) [inline src/Lean/Parser/Types.lean:311]: Drop-guided reuse analysis sure would be nice :smile:

**Comment by Kha** [comment]
!bench

File: src/Lean/Parser/Types.lean [hunk 2]
```diff
@@ -365,7 +365,11 @@ def adaptUncacheableContextFn (f : ParserContextCore → ParserContextCore) (p :
 
 /--
 Run `p` and record result in parser cache for any further invocation with this `parserName`, parser context, and parser state.
-`p` cannot access syntax stack elements pushed before the invocation in order to make caching independent of parser history. -/
+`p` cannot access syntax stack elements pushed before the invocation in order to make caching independent of parser history.
+As this excludes trailing parsers from being cached, we also reset `lhsPrec`, which is not read but set by leading parsers, to 0
+in order to increase cache hits. Finally, `errorMsg` is also reset to `none` as a leading parser should not be called in the first
+place if there was an error.
```
Comment (Kha) [inline src/Lean/Parser/Types.lean:371]: Sounds reasonable? Test suite says "yes". All `ParserState` fields are accounted for now.

Comment (leodemoura) [reply src/Lean/Parser/Types.lean:371]: I think so. It is great you are documenting it.

**Pull request merged**